### PR TITLE
MAKE-1028: scripting environment

### PR DESCRIPTION
**Pull request by tychoish**
this is basically done; I'm a bit worried about interactions with remote managers, which is currently disabled but I think might not be.

also need to implement methods for rpc and what not.

### CODE REVIEW

**Comment by tychoish** [comment]
I feel pretty good about the current state of this now, I've added rest and RPC. I think I'll need help to get the ssh/shell/etc methods/helpers

File: options/scripting_python.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+// NewPythonScriptingEnvironmnet generates a ScriptingEnvironment
+// taking the arguments given for later use. Use this function for
+// simple cases when you do not need or want to set as many aspects of
+// the environment configuration.
+func NewPythonScriptingEnvironmnet(path, reqtxt string, packages ...string) ScriptingEnvironment {
+	return &ScriptingPython{
+		CachedDuration:        time.Hour,
+		HostPythonInterpreter: "python3",
```
Comment (bsamek) [inline options/scripting_python.go]: What about the legacy python case?

File: options/scripting_python.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	if opts.HostPythonInterpreter == "" {
+		opts.HostPythonInterpreter = "python3"
+	}
+
+	return nil
+}
+
+func (opts *ScriptingPython) ID() string {
```
Comment (bsamek) [inline options/scripting_python.go]: This is potentially racy. Someone else could call ID and write to the cache. Is that a concern?

File: scripting_python.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func (e *pythonEnvironment) Run(ctx context.Context, args []string) error {
+	return e.manager.CreateCommand(ctx).Add(append([]string{e.opts.Interpreter()}, args...)).Run(ctx)
+}
+
+func (e *pythonEnvironment) RunScript(ctx context.Context, script string) error {
+	scriptChecksum := fmt.Sprintf("%x", sha1.Sum([]byte(script)))
+	wo := options.WriteFile{
+		Path:    filepath.Join(e.opts.VirtualEnvPath, "tmp", strings.Join([]string{e.manager.ID(), scriptChecksum}, "-")+".lisp"),
```
Comment (bsamek) [inline scripting_python.go]: copy pasta lisp in python file

File: scripting_roswell.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+func (e *roswellEnvironment) Build(ctx context.Context, dir string, args []string) error {
+	return e.manager.CreateCommand(ctx).Directory(dir).Environment(e.opts.Environment).AddEnv("ROSWELL_HOME", e.opts.Path).
+		SetOutputOptions(e.opts.Output).Add(append([]string{e.opts.Interpreter(), "dump", "executable"}, args...)).Run(ctx)
+}
+
+func (e *roswellEnvironment) Cleanup(ctx context.Context) error {
+	switch mgr := e.manager.(type) {
+	case RemoteClient:
+		return errors.Wrapf(mgr.CreateCommand(ctx).SetOutputOptions(e.opts.Output).AppendArgs("rm", "-rf", e.opts.Path).Run(ctx),
+			"problem removing remote python environment '%s'", e.opts.Path)
```
Comment (bsamek) [inline scripting_roswell.go]: python things in lisp file

File: options/scripting_python.go (hunk 1 above)
Comment (tychoish) [reply options/scripting_python.go]: the constructors are only for default situations. you can always construct directly.

File: options/scripting_python.go (hunk 2 above)
Comment (tychoish) [reply options/scripting_python.go]: yeahhhhhhhhh, I'm not super worried about it. 

I think the hashes will be stable. I'm more worried about someone having an options struct, creating it, losing track of the hash, and then changing it, and creating a different Environment, which I don't think will be a problem in reasonable situations.

**Pull request merged**